pkg/health/deduplicate.go

```diff
@@ -5,75 +5,121 @@ import (
 	"sync"
 )
 
-var _ Provider = (*deduplicator)(nil)
+var _ Provider = (*Deduplicator)(nil)
 
-// deduplicator is a health check provider that deduplicates health check reports
+// Deduplicator is a health check provider that deduplicates health check reports
 // i.e. it only reports a health check if the status or attributes have changed
-type deduplicator struct {
-	seen     sync.Map
+type Deduplicator struct {
+	lock     sync.Mutex
+	records  map[Check]*record
 	provider Provider
 }
 
 type record struct {
 	attr map[string]string
-	err  *string
+	err  error
 }
 
 func newOKRecord(attrs []Attr) *record {
 	return newRecord(nil, attrs)
 }
 
 func newErrorRecord(err error, attrs []Attr) *record {
-	errTxt := err.Error()
-	return newRecord(&errTxt, attrs)
+	return newRecord(err, attrs)
 }
 
-func newRecord(err *string, attrs []Attr) *record {
+func newRecord(err error, attrs []Attr) *record {
 	r := &record{err: err, attr: make(map[string]string)}
 	for _, a := range attrs {
 		r.attr[a.Key] = a.Value
 	}
 	return r
 }
 
+func (r *record) Attr() []Attr {
+	attrs := make([]Attr, 0, len(r.attr))
+	for k, v := range r.attr {
+		attrs = append(attrs, Attr{Key: k, Value: v})
+	}
+	return attrs
+}
+
 func (r *record) Equals(other *record) bool {
-	return r.equalError(other) &&
+	return equalError(r.err, other.err) &&
 		maps.Equal(r.attr, other.attr)
 }
 
-func (r *record) equalError(other *record) bool {
-	if r.err == nil || other.err == nil {
-		return r.err == other.err
+func equalError(a, b error) bool {
+	if a == nil || b == nil {
+		return a == b //nolint:errorlint
 	}
-	return *r.err == *other.err
+	return a.Error() == b.Error()
 }
 
-func NewDeduplicator(provider Provider) Provider {
-	return &deduplicator{provider: provider}
+func report(p Provider, check Check, err error, attrs ...Attr) {
+	if err != nil {
+		p.ReportError(check, err, attrs...)
+	} else {
+		p.ReportOK(check, attrs...)
+	}
 }
 
-func (d *deduplicator) swap(check Check, next *record) *record {
-	prev, there := d.seen.Swap(check, next)
-	if !there {
-		return nil
+func NewDeduplicator() *Deduplicator {
+	return &Deduplicator{
+		records:  make(map[Check]*record),
+		provider: &noopProvider{},
 	}
-	return prev.(*record)
 }
 
-// ReportError implements the Provider interface
-func (d *deduplicator) ReportError(check Check, err error, attrs ...Attr) {
-	cur := newErrorRecord(err, attrs)
-	prev := d.swap(check, cur)
-	if prev == nil || !cur.Equals(prev) {
-		d.provider.ReportError(check, err, attrs...)
+func (d *Deduplicator) SetProvider(p Provider) {
+	d.setProvider(p)()
+}
+
+func (d *Deduplicator) setProvider(p Provider) func() {
```

Comment:
What's the purpose of the closure function? It seems to always be invoked immediately.

Reply:
To be called after releasing the the lock. It's not very idiomatic, I updated implementation.